### PR TITLE
[Hub Generated] Review request for Microsoft.Consumption to add version preview/2019-04-01-preview

### PR DESCRIPTION
**Pull request by ampravinr**
If you are a MSFT employee you can view your [work branch via this link](https://portal.azure-devex-tools.com/app/branch/ampravinr/azure-rest-api-specs/usage-consumption-Microsoft.Consumption-2019-04-01-preview...master).

### Contribution checklist:
- [ ] I have reviewed the [documentation](https://github.com/Azure/azure-rest-api-specs#basics) for the workflow.
- [ ] [Validation tools](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/swagger-checklist.md#validation-tools-for-swagger-checklist) were run on swagger spec(s) and have all been fixed in this PR.

### CODE REVIEW

**Comment by adxsdknet** [comment]
# Automation for azure-sdk-for-net 
A PR has been created for you: 
https://github.com/azure/azure-sdk-for-net/pull/5778 
.NET SDK Commits: 
https://github.com/adxsdknet/azure-sdk-for-net/pull/5778/commits/a79e3e9fdbb1c582160823241387203cd4f7f295

File: specification/consumption/resource-manager/Microsoft.Consumption/preview/2019-04-01-preview/consumption.json
Comment (nschonni) [inline specification/consumption/resource-manager/Microsoft.Consumption/preview/2019-04-01-preview/consumption.json]:
```suggestion
    "/{scope}/providers/Microsoft.Consumption/usageDetails/download": {
```

Comment (ampravinr) [reply specification/consumption/resource-manager/Microsoft.Consumption/preview/2019-04-01-preview/consumption.json]: Thanks Nick for pointing this out. I have made the changes.

**Comment by veronicagg** [comment]
Have you evaluated warnings reported at https://portal.azure-devex-tools.com/app/branch/ampravinr/azure-rest-api-specs/usage-consumption-Microsoft.Consumption-2019-04-01-preview...master/linter ?

**Review by veronicagg — CHANGES_REQUESTED** [review]
Please review https://dev.azure.com/azure-sdk/public/_build/results?buildId=17699&view=logs&jobId=8d2f9c49-cd83-5a9d-f3fe-2fd651afa742&taskId=43324f1f-6724-5d61-785f-f0ac9fd256a1&lineStart=82&lineEnd=83&colStart=1&colEnd=1
and https://dev.azure.com/azure-sdk/public/_build/results?buildId=17699&view=logs&jobId=f10f5372-40ed-54fc-caa4-b1cb5bd1252b&taskId=d3b2b841-a93b-5fa5-47c3-d8c4f6eae6e3&lineStart=14&lineEnd=15&colStart=1&colEnd=1

**Comment by ampravinr** [comment]
> Have you evaluated warnings reported at https://portal.azure-devex-tools.com/app/branch/ampravinr/azure-rest-api-specs/usage-consumption-Microsoft.Consumption-2019-04-01-preview...master/linter ?

Yes @veronicagg . This is all existing which we would like to keep as is.

**Comment by ampravinr** [comment]
> Please review https://dev.azure.com/azure-sdk/public/_build/results?buildId=17699&view=logs&jobId=8d2f9c49-cd83-5a9d-f3fe-2fd651afa742&taskId=43324f1f-6724-5d61-785f-f0ac9fd256a1&lineStart=82&lineEnd=83&colStart=1&colEnd=1
> and https://dev.azure.com/azure-sdk/public/_build/results?buildId=17699&view=logs&jobId=f10f5372-40ed-54fc-caa4-b1cb5bd1252b&taskId=d3b2b841-a93b-5fa5-47c3-d8c4f6eae6e3&lineStart=14&lineEnd=15&colStart=1&colEnd=1

@veronicagg , i have reviewed and made the chages. Could you please verify from your side now.

**Comment by prem9in** [comment]
@flanakin fyi.

File: specification/consumption/resource-manager/Microsoft.Consumption/preview/2019-04-01-preview/consumption.json
Comment (RyanBensonMSFT) [inline specification/consumption/resource-manager/Microsoft.Consumption/preview/2019-04-01-preview/consumption.json]: This URL references an instance of usageDetails named download. This cannot be used the way you're intending. I think what you're trying to do here is download the usageDetail collection but there is no pattern in ARM for /actions on collections, only /actions on instances.
 
We can either include the download url in the response to usageDetails or add POST /{scope}/providers/Microsoft.Consumption/downloadUsageDetails

File: specification/consumption/resource-manager/Microsoft.Consumption/preview/2019-04-01-preview/consumption.json
Comment (RyanBensonMSFT) [inline specification/consumption/resource-manager/Microsoft.Consumption/preview/2019-04-01-preview/consumption.json]: There are very few commonalities between this model and the one defined in 2019-01-01. Are you depricating the old model or just introducing a large change through api-version?

File: specification/consumption/resource-manager/Microsoft.Consumption/preview/2019-04-01-preview/consumption.json
Comment (RyanBensonMSFT) [inline specification/consumption/resource-manager/Microsoft.Consumption/preview/2019-04-01-preview/consumption.json]: Looks like the following was removed:
          "ChargesListForEnrollmentAccount": {
            "$ref": "./examples/ChargesListForEnrollmentAccountFilterByStartEndDate.json"
          },
          "ChargesListByDepartment": {
            "$ref": "./examples/ChargesListForDepartmentFilterByStartEndDate.json"
          }

File: specification/consumption/resource-manager/Microsoft.Consumption/preview/2019-04-01-preview/consumption.json
Comment (RyanBensonMSFT) [inline specification/consumption/resource-manager/Microsoft.Consumption/preview/2019-04-01-preview/consumption.json]: Is this the replacement for $apply?

**Review by ryansbenson — COMMENTED** [review]
I'm having a hard time following the intent of this change as its vastly different than the previous GA'd api version.

File: specification/consumption/resource-manager/Microsoft.Consumption/preview/2019-04-01-preview/consumption.json
Comment (ampravinr) [reply specification/consumption/resource-manager/Microsoft.Consumption/preview/2019-04-01-preview/consumption.json]: Yes, this is intended. From this version onwards we are not supporting the Group by for this API. So removing the samples.

File: specification/consumption/resource-manager/Microsoft.Consumption/preview/2019-04-01-preview/consumption.json
Comment (ampravinr) [reply specification/consumption/resource-manager/Microsoft.Consumption/preview/2019-04-01-preview/consumption.json]: This will be a breaking change to us since Ibizza is already using this (internally) and they are consuming the downloadUrl. We can change it may be in a later version with backward compatible. But for this version we would like to keep it like downloadUrl. Also this is the same model we are keeping in our 2018-11-01-preview pricesheets downlaod as well.
- https://github.com/Azure/azure-rest-api-specs/blob/0972d9b7ba5b90b1c730ae744bcc655414d825bb/specification/consumption/resource-manager/Microsoft.Consumption/preview/2018-11-01-preview/consumption.json

File: specification/consumption/resource-manager/Microsoft.Consumption/preview/2019-04-01-preview/consumption.json
Comment (ampravinr) [reply specification/consumption/resource-manager/Microsoft.Consumption/preview/2019-04-01-preview/consumption.json]: its not replacement. We are introducing a new Query Param from this version onwards.

File: specification/consumption/resource-manager/Microsoft.Consumption/preview/2019-04-01-preview/consumption.json
Comment (ampravinr) [reply specification/consumption/resource-manager/Microsoft.Consumption/preview/2019-04-01-preview/consumption.json]: Yup the Charges API returns only single object (not array/list) for the API and tool was throwing error for the wrong samples. So I removed them and keeping only the right samples.

https://dev.azure.com/azure-sdk/public/_build/results?buildId=17699&view=logs&jobId=8d2f9c49-cd83-5a9d-f3fe-2fd651afa742&taskId=43324f1f-6724-5d61-785f-f0ac9fd256a1&lineStart=82&lineEnd=83&colStart=1&colEnd=1

File: specification/consumption/resource-manager/Microsoft.Consumption/preview/2019-04-01-preview/consumption.json
Comment (ampravinr) [reply specification/consumption/resource-manager/Microsoft.Consumption/preview/2019-04-01-preview/consumption.json]: Yes that's true. We are removing them from this versions and the previous versions still supports them.

File: specification/consumption/resource-manager/Microsoft.Consumption/preview/2019-04-01-preview/consumption.json
Comment (flanakin) [reply specification/consumption/resource-manager/Microsoft.Consumption/preview/2019-04-01-preview/consumption.json]: Let's please keep it as-is. We're actively working on a new API to replace this and can align with ARM async download standards as part of that. That said, is there a defined standard? I've seen different patterns from different RPs. I'd hate to align to one nonstandard format only to be asked to align to another a few months later.

File: specification/consumption/resource-manager/Microsoft.Consumption/preview/2019-04-01-preview/consumption.json
Comment (flanakin) [reply specification/consumption/resource-manager/Microsoft.Consumption/preview/2019-04-01-preview/consumption.json]: Fwiw, I had the same thought when I first saw this. That said, this already exists today on older API versions. Instead of changing this only to change it again in a few months, let's please keep this one as-is and fix this when we add download support to Microsoft.CostManagement/query.

File: specification/consumption/resource-manager/Microsoft.Consumption/preview/2019-04-01-preview/consumption.json
Comment (flanakin) [reply specification/consumption/resource-manager/Microsoft.Consumption/preview/2019-04-01-preview/consumption.json]: This changes the underlying metric we use to get data. The user has to pick which metric they want and get that data set. There isn't an option to get multiple, so it's not a $filter. (We thought about that and tried to start there, but it forced us to add more data and support a scenario we don't actually want to support, increasing surface area and introducing perf/scale risks due to the size of the request.)'

File: specification/consumption/resource-manager/Microsoft.Consumption/preview/2019-04-01-preview/consumption.json
Comment (ampravinr) [reply specification/consumption/resource-manager/Microsoft.Consumption/preview/2019-04-01-preview/consumption.json]: We are following the Async-operation approach for the downloads. As per the ARM stardards I believe this is the approach where we will mention the status url here in the AsyncOperation and the results url in the location property.

**Review by veronicagg — CHANGES_REQUESTED** [review]
Couple of inline questions.

File: specification/consumption/resource-manager/Microsoft.Consumption/preview/2019-04-01-preview/consumption.json
Comment (flanakin) [reply specification/consumption/resource-manager/Microsoft.Consumption/preview/2019-04-01-preview/consumption.json]: The intent is to make it very clear this is about downloading a file as opposed to a typical JSON GET request.

**Comment by veronicagg** [comment]
@ampravinr will you be making any changes based on the comments? are you following up with ARM team offline on their questions? Thanks!

**Comment by ampravinr** [comment]
Yes. Currently I am working to accommodate the Enum change you suggested on the query string  to make it Enum. Once I make changes and pushed the changes I will try to resume the Arm review to see review it and get sign off.
Also we where trying to see whether we can accommodate the change Arm suggested for listing different models based on kind and for now we cannot accommodate that so will go ahead with this for the Preview version.
Thanks,
Praveen Radhakrishnan
________________________________

File: specification/consumption/resource-manager/Microsoft.Consumption/preview/2019-04-01-preview/consumption.json
Comment (ampravinr) [reply specification/consumption/resource-manager/Microsoft.Consumption/preview/2019-04-01-preview/consumption.json]: I have made the changes. Could you please review and let me know if any issues.

File: specification/consumption/resource-manager/Microsoft.Consumption/preview/2019-04-01-preview/consumption.json
Comment (RyanBensonMSFT) [reply specification/consumption/resource-manager/Microsoft.Consumption/preview/2019-04-01-preview/consumption.json]: I don't know if following bad patterns is a great excuse. This will flag your RP as being non compliant with the RPC.

https://armwiki.azurewebsites.net/rp_onboarding/RPCstrongvalidation.html

Which API are you referring too?

**Comment by veronicagg** [comment]
@ampravinr ok, please work with ARM team on the ARM changes, so they sign off on the PR. 
Regarding CI I see python SDK for example failing to generate with this error: System.InvalidOperationException: Values specified by 'enum' mismatch those specified by 'x-ms-enum' (name: 'metrictype'): 'daily', 'monthly' vs 'usage', 'actualcost', 'amortizedcost'. Please take a [look](https://dev.azure.com/azure-sdk/public/_build/results?buildId=20062&view=logs&jobId=957f396e-b28a-53c4-b5bb-b8c6d959bd2d&taskId=142feabb-6d02-5968-caa1-f5b3c04c5cae&lineStart=134&lineEnd=134&colStart=41&colEnd=224).

File: specification/consumption/resource-manager/Microsoft.Consumption/preview/2019-04-01-preview/consumption.json
Comment (ms-premp) [reply specification/consumption/resource-manager/Microsoft.Consumption/preview/2019-04-01-preview/consumption.json]: @RyanBensonMSFT I am not following this feedback . What exactly is getting non-compliant with RPC ?
I see your point that at present in ARM, looks like there is no pattern with actions on collections.
Does this mean, since a pattern does not exist, we should not introduce it all. May be there was never a need for it and now we have a scenario for such actions.
The standard pattern for OData for adding action on collections is https://docs.microsoft.com/en-us/aspnet/web-api/overview/odata-support-in-aspnet-web-api/odata-v3/odata-actions#binding-an-action-to-an-entity-set
This makes sense for our users and it's a lot simpler to understand APIs, when we stick to standards. 
Please let me know, if I am missing something.

**Comment by ampravinr** [comment]
> @ampravinr ok, please work with ARM team on the ARM changes, so they sign off on the PR.
> Regarding CI I see python SDK for example failing to generate with this error: System.InvalidOperationException: Values specified by 'enum' mismatch those specified by 'x-ms-enum' (name: 'metrictype'): 'daily', 'monthly' vs 'usage', 'actualcost', 'amortizedcost'. Please take a [look](https://dev.azure.com/azure-sdk/public/_build/results?buildId=20062&view=logs&jobId=957f396e-b28a-53c4-b5bb-b8c6d959bd2d&taskId=142feabb-6d02-5968-caa1-f5b3c04c5cae&lineStart=134&lineEnd=134&colStart=41&colEnd=224).

@veronicagg - thanks for pointing me the error. I have fixed it. Please review and let me know if any issues.

**Comment by veronicagg** [comment]
I'm good from my side, approval pending on ARM team's signoff. Thanks!

File: specification/consumption/resource-manager/Microsoft.Consumption/preview/2019-04-01-preview/consumption.json
Comment (ampravinr) [reply specification/consumption/resource-manager/Microsoft.Consumption/preview/2019-04-01-preview/consumption.json]: I have removed the addiotional property.

**Comment by ms-premp** [comment]
@ravbhatnagar we are waiting on ARM Feedback on this PR. Please have a look. Our public announcements for this preview version is at present waiting on this swagger changes.

**Comment by ravbhatnagar** [comment]
Two concerns around this were discussed - 1. RBAC 2. Activity Log + Alert story. With this kind of modeling of action over collection. ARM will not parse the authorization action correctly as the instance segment is missing. hence what gets put in activity log will be wrong. And similarly RBAC will not be over the action which is what this API is. But since this is tenant level, RP handles its own RBAC. And since its tenant level, such logs dont get shown in activity log today. So these concerns are mitigated. ARM should look at this pattern and see if it makes sense to support it. 
The RP has agreed that this approval is for preview only. We will have another discussion for the GA story.
Signing off from ARM side.

**Comment by ampravinr** [comment]
@veronicagg - we have the Arm sign off. Could you please merge these changes.

**Review by veronicagg — APPROVED** [review]

**Comment by ampravinr** [comment]
Thanks Veronica for all the help on this.

Thanks,
Praveen Radhakrishnan

**Pull request merged**
